src/lib/communication/DY.Lib.Communication.Core.fst

```diff
@@ -0,0 +1,278 @@
+module DY.Lib.Communication.Core
+
+open Comparse
+open DY.Core
+open DY.Lib.State.PKI
+open DY.Lib.State.PrivateKeys
+open DY.Lib.Event.Typed
+
+#set-options "--fuel 0 --ifuel 0 --z3cliopt 'smt.qi.eager_threshold=100'"
+
+(*** Messages ***)
+
+[@@with_bytes bytes]
+type communication_message = {
+  sender:principal;
+  receiver:principal;  
+  payload:bytes;
+}
+
+%splice [ps_communication_message] (gen_parser (`communication_message))
+%splice [ps_communication_message_is_well_formed] (gen_is_well_formed_lemma (`communication_message))
+
+[@@with_bytes bytes]
+type signature_input = 
+  | Plain: msg:communication_message -> signature_input
+  | Encrypted: pk:bytes -> msg:communication_message -> signature_input
+
+#push-options "--ifuel 1 --fuel 0"
+%splice [ps_signature_input] (gen_parser (`signature_input))
+%splice [ps_signature_input_is_well_formed] (gen_is_well_formed_lemma (`signature_input))
+#pop-options
+
+instance parseable_serializeable_bytes_signature_input: parseable_serializeable bytes signature_input
+  = mk_parseable_serializeable ps_signature_input
+
+[@@with_bytes bytes]
+type communication_message_sign = {
```

Comment:
Nit: I would prefer `signed_communication_message` or something similar as the type name.